go-controller/pkg/ovn/controller/network_qos/network_qos_namespace.go

```diff
@@ -0,0 +1,135 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+)
+
+func (c *Controller) processNextNQOSNamespaceWorkItem(wg *sync.WaitGroup) bool {
+	wg.Add(1)
+	defer wg.Done()
+	nqosNSKey, quit := c.nqosNamespaceQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.nqosNamespaceQueue.Done(nqosNSKey)
+
+	err := c.syncNetworkQoSNamespace(nqosNSKey)
+	if err == nil {
+		c.nqosNamespaceQueue.Forget(nqosNSKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", nqosNSKey, err))
+
+	if c.nqosNamespaceQueue.NumRequeues(nqosNSKey) < maxRetries {
+		c.nqosNamespaceQueue.AddRateLimited(nqosNSKey)
+		return true
+	}
+
+	c.nqosNamespaceQueue.Forget(nqosNSKey)
+	return true
+}
+
+// syncNetworkQoSNamespace decides the main logic everytime
+// we dequeue a key from the nqosNamespaceQueue cache
+func (c *Controller) syncNetworkQoSNamespace(key string) error {
+	startTime := time.Now()
+	klog.V(5).Infof("Processing sync for Namespace %s in Network QoS controller", key)
+	defer func() {
+		klog.V(5).Infof("Finished syncing Namespace %s Network QoS controller: took %v", key, time.Since(startTime))
+	}()
+	namespace, err := c.nqosNamespaceLister.Get(key)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	// (i) namespace add
+	// (ii) namespace update because namespace's labels changed
+	// (iii) namespace delete
+	// case (iii)
+	if namespace == nil {
+		for _, cachedKey := range c.nqosCache.GetKeys() {
+			err := c.nqosCache.DoWithLock(cachedKey, func(nqosKey string) error {
+				if nqosObj, _ := c.nqosCache.Load(nqosKey); nqosObj != nil {
+					return c.clearNamespaceForNQOS(key, nqosObj)
+				}
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+		}
+		recordNamespaceReconcileDuration(c.controllerName, time.Since(startTime).Milliseconds())
+		return nil
+	}
+	// case (i)/(ii)
+	for _, cachedKey := range c.nqosCache.GetKeys() {
+		err := c.nqosCache.DoWithLock(cachedKey, func(nqosKey string) error {
+			if nqosObj, _ := c.nqosCache.Load(nqosKey); nqosObj != nil {
+				return c.setNamespaceForNQOS(namespace, nqosObj)
+			} else {
+				klog.Warningf("NetworkQoS not synced yet: %s", nqosKey)
+				// requeue nqos key to sync it
+				c.nqosQueue.Add(nqosKey)
+				// requeue namespace key 3 seconds later, allow NetworkQoS to be handled
+				c.nqosNamespaceQueue.AddAfter(key, 3*time.Second)
+				return nil
+			}
+		})
+		if err != nil {
+			return err
+		}
+	}
+	recordNamespaceReconcileDuration(c.controllerName, time.Since(startTime).Milliseconds())
+	return nil
+}
+
+// clearNamespaceForNQOS will handle the logic for figuring out if the provided namespace name
+// has pods that affect address sets of the cached network qoses. If so, remove them.
+func (c *Controller) clearNamespaceForNQOS(namespace string, nqosState *networkQoSState) error {
+	for _, rule := range nqosState.EgressRules {
+		if rule.Classifier == nil {
+			continue
```

Comment:
if classifier is nil, that means it covers all traffic, so in that case we have no address set?

Reply:
yes.